client/verta/verta/runtime.py

```diff
@@ -0,0 +1,93 @@
+# -*- coding: utf-8 -*-
+
+from collections import defaultdict
+import json
+import threading
+import os
+
+
+_THREAD = threading.local()
+
+def init_thread_context() -> None:
+    """Initialize our thread-local variable for storing logging context."""
+    if not hasattr(_THREAD, "context"):
+        _THREAD.context = defaultdict()
+
+
+def get_thread_context() -> defaultdict:
+    """
+    Return the current thread-local context or initialize a new one.
+    """
+    init_thread_context()
+    return _THREAD.context
+
+
+def set_thread_context(context: defaultdict) -> defaultdict:
+    """
+    Sets the thread-local context, overwriting any existing values.
+    """
+    # init_thread_context()
+    _THREAD.context = context
+    return _THREAD.context
+
+
+def log(**kwargs) -> None:
+    """
+    Log any number of key-value pairs to the current prediction context.
+    Context is stored in thread-local variables.
+
+    Parameters
+    ----------
+    kwargs : Any
+        Arguments and values to be added as key-value pairs to the current
+        logging context. Values must be serializable to JSON.
+
+    Returns
+    -------
+    None
+    """
+    local_context: defaultdict = get_thread_context().copy()
+    local_context.update(**kwargs)
+    set_thread_context(local_context)
+
+
+class context:
+    """
+    Context manager for aggregating key-value pairs into a custom log entry.
+    Uses thread-local variables to store context in a thread-safe manner.
+
+    Parameters
+    ----------
+    kwargs : Any
+        Key-value pairs to be added to the current logging context.
+    Attributes
+    ----------
+    as_json : str
+        A JSON string representation of all the current saved logging context.
+    """
+    def __init__(self, **kwargs):
+        self.new_context = kwargs
```

Comment:
I think I'm a bit confused by how this is to be used. Is

```python
with context(key=val) as ctx:
    pass
```

just equivalent to

```python
with context() as ctx:
    ctx.log(key=val)
```

?

Considering that `context` is meant to be instantiated either by our Python Model Service, or by the user trying to emulate the model service, I'm not sure that it makes sense to build logging functionality into `__init__()`

Reply:
I refactored this to simplify it.  I had originally thought nesting these context managers would be a thing, but we do not need that for this particular iteration.  We could add the ability to nest, and even wrap a logging.Logger class to mimic python native logging calls like `warning`, or `error` etc. if customers want a more sophisticated implementation in the future.